### DISCUSSION TITLE
Integration test suggestion?

### DISCUSSION THREAD

**Discussion by jcyh0120** [pocketbase/pocketbase#603] [Q&A] [answered]
I'm using pocketbase with a simple nodejs api server. Since pocketbase is being actively developed. I'd like to add some integration test in my project. However, the only way I can do is mock all the pocketbase js sdk in my jest test case now.

Is there any better suggestion to run pocketbase like sqlite with :memory config so the data would be seperated from each test case?

I really like the concept of pocketbase. Every decision made consist with the priciple of ease of use and simplicity.

**Comment by ganigeorgiev** [comment, marked as answer]
Hm, complete integration testing with an external backend is difficult to do. PocketBase provides some test helpers, but they are useful only when you embed PocketBase in your Go app - https://pocketbase.io/docs/testing/.

You can't start PocketBase with in memory sqlite, but even if that was possible you still will not have a separate db while running the same PocketBase instance. For full integration tests, you'll need not only the database but the file storage abstraction and therefore it may be easier to just start PocketBase with an entire `test_pb_data` directory:

```sh
./pocketbase serve --dir="/path/to/test_pb_data/"
```

The internal integration tests uses similar approach, but since we have access to the Go code, we can easily copy the test data to a new location and run the tests against the temporary copy without interfering with the other tests.

With nodejs I'm not sure how this could be achieved. I would try the following:

1. Create your test mock data by starting PocketBase with `./pocketbase serve --dir="/path/to/test_pb_data/"`. Once done, terminate the process. `test_pb_data` now contains your test data that you'll use with each test run.

2. You can start new PocketBase instance with a copy of your test data on `beforeEach` and clean up the temp data and process on `afterEach`. Something like this (**the below is just an example, I haven't tested it!**):

```js
const { execSync } = require("node:child_process");

// you can use any other library for copying directories recursively
const fse = require("fs-extra");

let controller; // this will be used to terminate the PocketBase process

const srcTestDirPath = "/path/to/test_pb_data";
const tempTestDirPath = "/path/to/temp_test_pb_data";

beforeEach(() => {
  // copy test_pb_data to a temp location
  fse.copySync(srcTestDirPath, tempTestDirPath);

  controller = new AbortController();

  // start PocketBase with the test_pb_data
  execSync("./pocketbase serve --dir=" + tempTestDirPath, { signal: controller.signal});
});

afterEach(() => {
  // stop the PocketBase process
  controller.abort();

  // clean up the temp test directory
  fse.removeSync(tempTestDirPath);
});

test('test against an actual PocketBase server running on localhost:8090', () => {
  ...
});
```

The above would work only if your tests are executed synchronously (for parallel execution, you'll need additional setup to switch between the different instances). But PocketBase startup time is almost instantaneous, so it shouldn't affect too much your tests performance.

**Reply by jcyh0120** [reply]
Thanks! Super helpful for beginner of pocketbase with these instructions. It’s quite enough for play around, dev and test. Appreciate for the detail example above.👍

**Reply by parkernilson** [reply]
Has anybody gotten this type of solution to work? I have tried using docker compose to serve a container running pocketbase, however I keep ending up with a race condition where my tests run before pocketbase is ready. Is there a way to wait until pocketbase is ready?

**Reply by ErtanTaner** [reply]
Hi @parkernilson, you can wait for test with https://www.npmjs.com/package/wait-on package. Wait-on can listen for a spesific port is available or not. Just listen for the port which pocketbase going to serve and run tests after that.
